README.md

```diff
@@ -16,3 +16,142 @@ bower install purescript-argonaut-core
 ## Documentation
 
 Module documentation is [published on Pursuit](http://pursuit.purescript.org/packages/purescript-argonaut-core).
+
+## Tutorial
+
+Some of Argonaut's functions might seem a bit arcane at first, so it can help
+to understand the underlying design decisions which make it the way it is.
+
+One approach for modelling JSON values would be to define an ADT, like this:
+
+```purescript
+data Json
+  = JNull
+  | JBoolean Boolean
+  | JArray (Array Json)
+  | JObject (StrMap Json)
+  [...]
+```
+
+And indeed, some might even say this is the obvious approach.
+
+Because Argonaut is written with the compilation target of JavaScript in mind,
+it takes a slightly different approach, which is to reuse the existing data
+types which JavaScript already provides. This way, the result of JavaScript's
+`JSON.stringify` function is already a `Json` value, and no extra processing is
+needed before you can start operating on it. This ought to help your program
+both in terms of speed and memory churn.
+
+Much of the design of Argonaut follows naturally from this design decision.
+
+### Introducing Json values
+
+(Or, where do `Json` values come from?)
+
+Usually, a `Json` value will be introduced into your program via either the FFI
+or via the construction functions in `Data.Argonaut.Core`. Here are some
+examples:
+
+```javascript
+// In an FFI module.
+exports.someNumber = 23.6;
+exports.someBoolean = false;
+exports.someObject = {people: [{name: "john"}, {name: "jane"}], 
+```
+
+```purescript
+foreign import someNumber :: Json
+foreign import someBoolean :: Json
+foreign import someObject :: Json
+```
+
+Generally, if a JavaScript value could be returned from a call to
+`JSON.stringify`, it's fine to import it from the FFI as `Json`. So, for
+example, objects, booleans, numbers, strings, and arrays are all fine, but
+functions are not.
+
+The construction functions (that is, `fromX`, or `jsonX`) can be used as
+follows:
+
+```purescript
+import Data.Tuple (Tuple(..))
+import Data.StrMap as StrMap
+import Data.Argonaut.Core as A
+
+someNumber = A.fromNumber 23.6
+someBoolean = A.fromBoolean false
+someObject = A.fromObject (StrMap.fromFoldable [
+                Tuple "people" (A.fromArray [
+                  A.jsonSingletonObject "name" (A.fromString "john"),
+                  A.jsonSingletonObject "name" (A.fromString "jane")
+                ])
+              ])
+```
+
+### Eliminating/matching on `Json` values
```

Comment:
hmm - the following section I found a bit tricky.

It absolutely helped me (at least I think so) to understand why I would use `foldJSON`s but its not really clear from the text at the moment.

My suggestion is to restructure it as follows

- explain the `toX`methods and that its usually not enough as you only "convert" very superficially. for example `toArray` yields an `Maybe (Array Json)`
- thats why you will need the `foldJsonX` functions
- start with examples the single ones and examples  
        `fromJsonString = foldJsonString "" \x -> x` (btw. I would the more elaborate syntax in examples. For example I had to look up that `_` can be used for partial application. That was new to me)  
then show a more useful example of a fold like `isLolString``
- finally move to the fullblown `foldJson`
- and at last I would do complete example that shows how to fold a particular Json to a defined type.

    data Bar = Bar (Array Int)
    data SomeObject = SomeObject {foo:: Int, bar:: Bar}

    defaultSomeObject = SomeObject {foo: 1, Bar []}
    toSomeObject = foldJson defaultSomeObject \x -> ...

    so = toSomeObject someObjectFromFFI

I hope my thoughts are correct and make sense.